### PR TITLE
Dockerfile - Update CUDA 11.1.1 Dockerfile

### PR DESCRIPTION
**Pull request by abuccts**
__Description__

Update packages and add build cache for CUDA 11.1.1 Dockerfile.

__Major Revisions__

* remove duplicate cmake and ompi, which are already in base image
* add hpcx and sharp lib
* add cache for gitmodules build

### CODE REVIEW

File: third_party/Makefile [hunk 1]
```diff
@@ -14,5 +14,5 @@ cutlass:
 ifneq (,$(wildcard cutlass/CMakeLists.txt))
 	cmake -DCMAKE_INSTALL_BINDIR=$(SB_MICRO_PATH)/bin -DCMAKE_INSTALL_LIBDIR=$(SB_MICRO_PATH)/lib -DCMAKE_BUILD_TYPE=Release \
 		-DCUTLASS_NVCC_ARCHS='70;80' -DCUTLASS_ENABLE_EXAMPLES=OFF -DCUTLASS_ENABLE_TESTS=OFF -S ./cutlass -B ./cutlass/build
-	cmake --build ./cutlass/build -j 8 --target install
+	cmake --build ./cutlass/build --target install
```
Comment (cp5555) [inline third_party/Makefile:17]: `-j 8` doesn't work here?

File: dockerfile/cuda11.1.1.dockerfile [hunk 2]
```diff
@@ -42,39 +35,47 @@ RUN mkdir -p /root/.ssh && \
...
     make -j src.build && \
     make install
+
+# TODO: move to gitmodules
+RUN git clone -b v4.5-0.2 https://github.com/linux-rdma/perftest.git /usr/local/perftest && \
+    cd /usr/local/perftest && \
+    ./autogen.sh && \
+    ./configure CUDA_H_PATH=/usr/local/cuda/include/cuda.h && \
+    make -j && \
+    make install
```
Comment (cp5555) [inline dockerfile/cuda11.1.1.dockerfile:78]: What's the purpose for this?

File: dockerfile/cuda11.1.1.dockerfile [hunk 3]
```diff
@@ -42,39 +35,47 @@ RUN mkdir -p /root/.ssh && \
...
     make -j && \
-    make install
+    make install && \
+    cd /tmp && \
+    rm -rf nccl-rdma-sharp-plugins
 
-# Install NCCL
-RUN git clone -b v2.8.4-1 https://github.com/NVIDIA/nccl /usr/local/nccl && \
+# Install NCCL patch
+RUN git clone -b bootstrap_tag https://github.com/NVIDIA/nccl.git /usr/local/nccl && \
```
Comment (cp5555) [inline dockerfile/cuda11.1.1.dockerfile:67]: Shall we upgrade to 2.9.9 directly?

File: dockerfile/cuda11.1.1.dockerfile [hunk 4]
```diff
@@ -42,39 +35,47 @@ RUN mkdir -p /root/.ssh && \
...
-    cd openmpi-${OMPI_VERSION} && \
-    ./configure --enable-orterun-prefix-by-default && \
-    make -j all && \
-    make install && \
-    ldconfig && \
-    rm -rf /tmp/openmpi-${OMPI_VERSION}*
-
 # Install OFED
-ARG OFED_VERSION=5.2-1.0.4.0
+ENV OFED_VERSION=5.2-2.2.3.0
```
Comment (cp5555) [inline dockerfile/cuda11.1.1.dockerfile:39]: Shall we check the version?

File: dockerfile/cuda11.1.1.dockerfile [hunk 5]
```diff
@@ -42,39 +35,47 @@ RUN mkdir -p /root/.ssh && \
...
-RUN cd /tmp && \
-    wget -q https://www.open-mpi.org/software/ompi/v4.0/downloads/openmpi-${OMPI_VERSION}.tar.gz && \
-    tar xzf openmpi-${OMPI_VERSION}.tar.gz && \
-    cd openmpi-${OMPI_VERSION} && \
-    ./configure --enable-orterun-prefix-by-default && \
-    make -j all && \
-    make install && \
-    ldconfig && \
-    rm -rf /tmp/openmpi-${OMPI_VERSION}*
-
```
Comment (cp5555) [inline dockerfile/cuda11.1.1.dockerfile:56]: Why remove OpenMPI?

File: dockerfile/cuda11.1.1.dockerfile [hunk 6]
```diff
@@ -24,15 +24,8 @@ RUN apt-get update && \
...
     libcap2 \
-    libtinfo5
-
-# Install CMake
-RUN wget -q https://github.com/Kitware/CMake/releases/download/v3.17.1/cmake-3.17.1-Linux-x86_64.sh \
-    -O /tmp/cmake-install.sh && \
-    chmod +x /tmp/cmake-install.sh && \
-    mkdir /usr/local/cmake && \
-    /tmp/cmake-install.sh --skip-license --prefix=/usr/local/cmake && \
-    rm /tmp/cmake-install.sh
```
Comment (cp5555) [inline dockerfile/cuda11.1.1.dockerfile:35]: Why remove CMake? Would you please write the reason in description?

File: third_party/Makefile (hunk 1 above)
Comment (abuccts) [reply third_party/Makefile:17]: it's too slow on test node, added back

File: dockerfile/cuda11.1.1.dockerfile (hunk 2 above)
Comment (abuccts) [reply dockerfile/cuda11.1.1.dockerfile:78]: perftest?

File: dockerfile/cuda11.1.1.dockerfile (hunk 3 above)
Comment (abuccts) [reply dockerfile/cuda11.1.1.dockerfile:67]: we should use nvidia images' built-in nccl, it has some issues in this 20.12 version so rebuild with patch.
we could provide another 21.05 version which has nccl 2.9.9

File: dockerfile/cuda11.1.1.dockerfile (hunk 4 above)
Comment (abuccts) [reply dockerfile/cuda11.1.1.dockerfile:39]: it's the same with hpcx package

File: dockerfile/cuda11.1.1.dockerfile (hunk 5 above)
Comment (abuccts) [reply dockerfile/cuda11.1.1.dockerfile:56]: > remove duplicate cmake and ompi

it's already in description

File: dockerfile/cuda11.1.1.dockerfile (hunk 6 above)
Comment (abuccts) [reply dockerfile/cuda11.1.1.dockerfile:35]: > remove duplicate cmake and ompi

it's already in description

Comment (cp5555) [reply dockerfile/cuda11.1.1.dockerfile:35]: I mean the reason.

File: dockerfile/cuda11.1.1.dockerfile (hunk 5 above)
Comment (cp5555) [reply dockerfile/cuda11.1.1.dockerfile:56]: I mean the reason.

File: dockerfile/cuda11.1.1.dockerfile (hunk 2 above)
Comment (cp5555) [reply dockerfile/cuda11.1.1.dockerfile:83]: Yes.

File: dockerfile/cuda11.1.1.dockerfile (hunk 6 above)
Comment (abuccts) [reply dockerfile/cuda11.1.1.dockerfile:35]: it's already in base image and duplicate here, updated the description

File: dockerfile/cuda11.1.1.dockerfile (hunk 5 above)
Comment (abuccts) [reply dockerfile/cuda11.1.1.dockerfile:56]: it's already in base image and duplicate here, updated the description

File: dockerfile/cuda11.1.1.dockerfile (hunk 2 above)
Comment (abuccts) [reply dockerfile/cuda11.1.1.dockerfile:83]: it includes cuda aware ib_write_bw etc., it was built here for quick debug

**Review by cp5555 — APPROVED** [review]

File: dockerfile/cuda11.1.1.dockerfile [hunk 7]
```diff
@@ -2,7 +2,7 @@ FROM nvcr.io/nvidia/pytorch:20.12-py3
```
Comment (cp5555) [inline dockerfile/cuda11.1.1.dockerfile:2]: Add comments for all version information.

**Review by guoshzhao — APPROVED** [review]

File: dockerfile/cuda11.1.1.dockerfile (hunk 7 above)
Comment (abuccts) [reply dockerfile/cuda11.1.1.dockerfile:2]: added

**Pull request merged**